### PR TITLE
Stats step 1: stats v4 UI

### LINKED ISSUE #1067: Stats Step 1: add UI to support the new stats v4 api
Use the feature flag to determine if the user has the new stats feature enabled. (For now this means only devs).

Display the new Stats for Today, This Week, This Month, and This Year

stats-improvements-2019-05-17-00-20-02

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #1067 
Based on https://github.com/woocommerce/woocommerce-ios/pull/1184

- [x] ⚠️ Update PR base to `develop` before merging ⚠️ 

## Changes
- Sorry about the size of this PR, most code are view controllers (and a xib from v3)
- Reset stats v4 data on switching store
- Added `chartAxisHourFormatter: DateFormatter` to show hour date format on stats v4 "today" chart x-axis
- The changes in `UIView+Helpers` are intended to be in `WordPressUI` and currently pending code review: https://github.com/wordpress-mobile/WordPressUI-iOS/pull/39
- In `DashboardUIFactory`, showed stats v4 UI if building from Xcode
- In `PeriodDataViewController` and `TopPerformerDataViewController`, moved protocol conformance to the extension that implements the `IndicatorInfoProvider` protocol
- (The post about stats v4 UI hierarchy would be helpful for reviewing the view controllers) Top container view controller `StoreStatsAndTopPerformersViewController`:
  - `StoreStatsAndTopPerformersViewController` is a subclass of `ButtonBarPagerTabStripViewController`, and a lot of button tab bar setup is the same as in v3 `StoreStatsViewController` and `TopPerformersViewController`
  - 4 child view controllers `StoreStatsAndTopPerformersViewController` for each time range tab:
    - `StoreStatsAndTopPerformersViewController` is a container of scrollable stack view of:
      - Store stats data view, managed by child view controller `StoreStatsV4PeriodViewController`
        - `StoreStatsV4PeriodViewController` is mostly the same as v3 `PeriodDataViewController` (and `PeriodDataViewController.xib`) with time range instead of granularity enum and different Fetched Results Controller setup for store stats and site visit stats
      - Top performers header view `TopPerformersSectionHeaderView` (created in this PR)
      - Top performers data view, managed by child view controller `TopPerformerDataViewController` (same in v3)

Stats UI work

## Testing
Prerequisite: the site has WC admin extension installed and activated
- Build this branch and run from Xcode for stats v4 UI
- Go to "My store" tab
- Check each time range tab: "Today"/"This Week"/"This Month"/"This Year" --> the stats should be consistent with the WC admin dashboard on `/wp-admin`

## Example screenshots

\ | Today | This week | This month | This year
-- | -- | -- | -- | --
Store stats | Simulator Screen Shot - iPhone 5s - 2019-08-15 at 17 19 49 | Simulator Screen Shot - iPhone 5s - 2019-08-15 at 17 19 56 | Simulator Screen Shot - iPhone 5s - 2019-08-15 at 17 20 03 | Simulator Screen Shot - iPhone 5s - 2019-08-15 at 17 20 08
Top performers | Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 37 48 | Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 37 51 | Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 49 59 | Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 49 45

Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

**Comment by mindgraffiti** [comment]
I was reviewing the code and realized, I'm not sure what type of testing do you run? A store with v3 stats and a store with v4? Any other test cases I should try out?

**Comment by jaclync** [comment]
> I was reviewing the code and realized, I'm not sure what type of testing do you run? A store with v3 stats and a store with v4? Any other test cases I should try out?

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+        scrollView.refreshControl = refreshControl
+
+        scrollView.addSubview(stackView)
+        scrollView.translatesAutoresizingMaskIntoConstraints = false
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        scrollView.pinSubviewToAllEdges(stackView)
+        NSLayoutConstraint.activate([
+            stackView.widthAnchor.constraint(equalTo: scrollView.widthAnchor)
+            ])
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift]: Closing parens tabbed over one too many

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        childViewContrllers.forEach { childViewController in
+            childViewController.view.translatesAutoresizingMaskIntoConstraints = false
+        }
+
+        // Store stats.
+        let storeStatsPeriodView = storeStatsPeriodViewController.view!
+        stackView.addArrangedSubview(storeStatsPeriodView)
+        NSLayoutConstraint.activate([
+            storeStatsPeriodView.heightAnchor.constraint(equalToConstant: 380),
+            ])
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift]: same here

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            topPerformersHeaderView.heightAnchor.constraint(equalToConstant: 44),
+            // Top border view
+            headerTopBorderView.topAnchor.constraint(equalTo: topPerformersHeaderView.topAnchor),
+            headerTopBorderView.leadingAnchor.constraint(equalTo: topPerformersHeaderView.leadingAnchor),
+            headerTopBorderView.trailingAnchor.constraint(equalTo: topPerformersHeaderView.trailingAnchor),
+            // Bottom border view
+            headerBottomBorderView.bottomAnchor.constraint(equalTo: topPerformersHeaderView.bottomAnchor),
+            headerBottomBorderView.leadingAnchor.constraint(equalTo: topPerformersHeaderView.leadingAnchor),
+            headerBottomBorderView.trailingAnchor.constraint(equalTo: topPerformersHeaderView.trailingAnchor),
+            ])
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift]: and here

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+        // Top performers.
+        let topPerformersPeriodView = topPerformersPeriodViewController.view!
+        stackView.addArrangedSubview(topPerformersPeriodView)
+        NSLayoutConstraint.activate([
+            topPerformersPeriodView.heightAnchor.constraint(equalToConstant: 359.5),
+            topPerformersPeriodView.heightAnchor.constraint(greaterThanOrEqualToConstant: 359.5)
+            ])
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift]: ditto

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            headerBottomBorderView.bottomAnchor.constraint(equalTo: topPerformersHeaderView.bottomAnchor),
+            headerBottomBorderView.leadingAnchor.constraint(equalTo: topPerformersHeaderView.leadingAnchor),
+            headerBottomBorderView.trailingAnchor.constraint(equalTo: topPerformersHeaderView.trailingAnchor),
+            ])
+
+        // Top performers.
+        let topPerformersPeriodView = topPerformersPeriodViewController.view!
+        stackView.addArrangedSubview(topPerformersPeriodView)
+        NSLayoutConstraint.activate([
+            topPerformersPeriodView.heightAnchor.constraint(equalToConstant: 359.5),
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift]: why `.5` in this constraint? Is this to account for the border line getting drawn in?

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        }
+    }
+
+    func createBorderView() -> UIView {
+        let view = UIView(frame: .zero)
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.backgroundColor = StyleManager.wooGreyBorder
+        NSLayoutConstraint.activate([
+            view.heightAnchor.constraint(equalToConstant: 1)
+            ])
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift]: ditto

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+// MARK: - User Interface Configuration
+//
+private extension StoreStatsAndTopPerformersViewController {
+    func createBorderView() -> UIView {
+        let view = UIView(frame: .zero)
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.backgroundColor = StyleManager.wooGreyBorder
+        NSLayoutConstraint.activate([
+            view.heightAnchor.constraint(equalToConstant: 1)
+            ])
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift]: Closing parens tabbed over one too many

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift [hunk 8]
```diff
@@ -0,0 +1,609 @@
...
+    ///
+    private lazy var orderStatsResultsController: ResultsController<StorageOrderStatsV4> = {
+        let storageManager = AppDelegate.shared.storageManager
+        let predicate = NSPredicate(format: "timeRange ==[c] %@", timeRange.rawValue)
+        return ResultsController(storageManager: storageManager, matching: predicate, sortedBy: [])
+    }()
+
+    /// Placeholder: Mockup Charts View
+    ///
+    private lazy var placehoderChartsView: ChartPlaceholderView = ChartPlaceholderView.instantiateFromNib()
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift]:
```suggestion
    private lazy var placeholderChartsView: ChartPlaceholderView = ChartPlaceholderView.instantiateFromNib()
```

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift [hunk 9]
```diff
@@ -0,0 +1,609 @@
...
+        let code = CurrencySettings.shared.currencyCode
+        return CurrencySettings.shared.symbol(from: code)
+    }
+
+    private var summaryDateUpdated: String {
+        if let lastUpdatedDate = lastUpdatedDate {
+            return lastUpdatedDate.relativelyFormattedUpdateString
+        } else {
+            return ""
+        }
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsV4PeriodViewController.swift]: This block can be re-written to remove the `else`:

```
if let lastUpdatedDate = lastUpdatedDate {
            return lastUpdatedDate.relativelyFormattedUpdateString
        } 

return ""
}
```

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift (hunk 5 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift]: I calculated the height from the v3 top performers height of 465px, deducting the following UI components that are in the vertical stack view (in `Dashboard.storyboard`):

header view (moved to a separate view `TopPerformersSectionHeaderView`): 55.5px
border: 1px
button bar view (moved to the top as the time range selector): 48px
border: 1px
——
105.5px

Then I got 359.5px for just the top performers data view

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift (hunk 1 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift]: hmm this is how Xcode formats the indentation though, if you move that line around or ctrl+i the whole code block

Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersPeriodViewController.swift]: maybe we can look into some auto-formatting tool like swift lint to detect/correct these formatting details for us?

**Comment by jaclync** [comment]
Addressed the top performers bottom border in https://github.com/woocommerce/woocommerce-ios/pull/1185/commits/455e1739ebd44d295878890e58d5757fe52266fe (also added a 44px empty view below it, the design spec has 181px spacing and I'll confirm with Kylea). Some screenshots:

no data view | 2 top performers
-- | --
Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 37 48 | Simulator Screen Shot - iPhone 5s - 2019-08-16 at 11 37 51

**Comment by jaclync** [comment]
Responded to the CR comments with some corresponding code updates in the last 3 commits. Merging now, lemme know if there is anything I missed and I'll address in a separate PR. Also let's revisit the [formatting topic](https://github.com/woocommerce/woocommerce-ios/pull/1185#discussion_r314569221) in p2!

**Pull request merged**